src/main/java/org/csanchez/jenkins/plugins/kubernetes/pod/retention/OnJobFailure.java

```diff
@@ -0,0 +1,224 @@
+
+import java.util.stream.Collectors;
+
+/**
+ * This pod retention policy keeps the pod from being terminated if the Jenkins
+ * job it's associated with fails.
+ *
+ * In case of any other result, including errors in determining the result, it
+ * will default to deleting the pod.
+ */
+public class OnJobFailure extends PodRetention implements Serializable {
+
+    private static final long serialVersionUID = -6422177946264212816L;
+
+    private static final Logger LOGGER = Logger.getLogger(OnJobFailure.class.getName());
+
+    private static final String MODULENAME = "OnJobFailure";
+
+    // small convenience function
+    private void LOG(Level level, String message) {
+        LOGGER.log(level, () -> MODULENAME + ": " + message);
+    }
+
+    @DataBoundConstructor
+    public OnJobFailure() {
+    }
+
+    @Override
+    public boolean shouldDeletePod(KubernetesCloud cloud, Pod pod) {
+        if (cloud == null || pod == null) {
+            LOG(Level.INFO, "shouldDeletePod called without actual cloud and pod");
+            return true;
+        }
+
+        // Get the current Jenkins instance to access a list of all jobs
+        Jenkins jenkins = Jenkins.getInstanceOrNull();
+        if (jenkins == null) {
+            LOG(Level.INFO, "Couldn't get the current Jenkins reference");
+            return true;
+        }
+
+        // All known jobs of the current Jenkins instance
+        List<Job> jobs = jenkins.getAllItems(Job.class);
+        if (jobs.isEmpty()) {
+            LOG(Level.INFO, "Jenkins doesn't have any jobs?");
+            return true;
+        }
+
+        // runUrl will be something like "job/<name>/<runId>/" or
+        // "job/<folder>/job/<name>/<runId>/" if nested
+        // this is the trick how we get our job name and run id
+        String runUrl = pod.getMetadata().getAnnotations().get("runUrl");
+        if (runUrl == null) {
+            LOG(Level.INFO, "The pod has no required 'runUrl' annotation");
+            return true;
+        }
+
+        // everything is in place, get the result
+        Result result = getResultForJob(runUrl, jobs);
+        if (result == null) {
+            // we couldn't get the result for some reason
+            LOG(Level.INFO, "Couldn't find the result for runUrl: " + runUrl);
+            return true;
+        }
+
+        // finally, delete only if successful
+        boolean delete = result.equals(Result.SUCCESS);
+        LOG(Level.FINE, "delete = " + delete);
+        return delete;
+    }
+
+    /**
+     * Split up the runUrl string and return the run id
+     *
+     * @param runUrl the "runUrl" annotation of the kubernetes pod
+     * @return the run id as a string
+     */
+    public String getRunId(String runUrl) {
+        // extract the relevant parts
+        String[] parts = runUrl.split("/");
+
+        if (parts.length < 3) {
+            LOG(Level.INFO, "runUrl has unknown format: " + runUrl);
+            return null;
+        }
```

Comment:
This is way too simplistic. If you need to look up the run, I would advise instead to add new annotations (one for the item full name, another for the build id)